### PR TITLE
Add CAPQ, upgrade SprayList, upgrade Linden and add new benchmark

### PR DESCRIPTION
**Pull request by kjellwinblad**
This pull request contains three commits:

1. The first commit adds the priority queues CA-PQ, CA-DM, CA-IN and CATree to the random benchmark. They are all described in the LCPC'2016 paper [The Contention Avoiding Concurrent Priority Queue][1].
2. The second commit upgrades both the SprayList implementation and the Linden priority queue implementation.
3. The last commit adds a new shortest paths benchmark that reads the input graph from a file.

Please, expand the commit messages below for details about these changes.

The benchmark code and the priority queues have been tested by running the new shortest paths with up to 64 worker threads on a machine with 64 hardware threads. To check that the priority queues and the benchmark work as intended, the calculated distances were compared against the actual distances after each run. 

[1]: https://www.it.uu.se/research/group/languages/software/ca_pq

### CODE REVIEW

File: bench.py [hunk 1]
```diff
@@ -5,7 +5,11 @@
 
 from optparse import OptionParser
 
-ALGORITHMS = [ 'cheap'
+ALGORITHMS = [ 'capq'
```
Comment (schuay) [inline bench.py]: Nit: Please alpha-sort these.

File: lib/capq/binheap_test.c [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with kpqueue.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdio.h>
+#include "binheap.h"
+
+int main()
```
Comment (schuay) [inline lib/capq/binheap_test.c:23]: Nit: Could add this to the test suite (or remove it).

File: src/bench/random.cpp [hunk 3]
```diff
@@ -31,6 +31,7 @@
 #include <valgrind/callgrind.h>
 #endif
 
+#include "pqs/cppcapq.h"
```
Comment (schuay) [inline src/bench/random.cpp]: Nit: Please alpha-sort pqs/.

File: src/bench/random.cpp [hunk 4]
```diff
@@ -188,15 +199,17 @@ usage()
...
             DEFAULT_SIZE,
             KEYS_UNIFORM, KEYS_ASCENDING, KEYS_DESCENDING, KEYS_RESTRICTED_8, KEYS_RESTRICTED_16, DEFAULT_KEYS,
             DEFAULT_NTHREADS,
             DEFAULT_SEED,
             WORKLOAD_UNIFORM, WORKLOAD_SPLIT, WORKLOAD_PRODUCER, WORKLOAD_ALTERNATING, DEFAULT_WORKLOAD,
-            PQ_CHEAP, PQ_DLSM, PQ_GLOBALLOCK, PQ_KLSM16, PQ_KLSM128, PQ_KLSM256, PQ_KLSM4096,
-            PQ_LINDEN, PQ_LSM, PQ_MLSM, PQ_MULTIQ, PQ_SEQUENCE, PQ_SKIP, PQ_SLSM, PQ_SPRAY);
+            PQ_CAPQ, PQ_CADM, PQ_CAIN, PQ_CATREE, PQ_CHEAP, PQ_DLSM, PQ_GLOBALLOCK, PQ_KLSM16,
+            PQ_KLSM128, PQ_KLSM256, PQ_KLSM4096, PQ_LINDEN, PQ_LSM, PQ_MLSM, PQ_MULTIQ, PQ_SEQUENCE,
+            PQ_SKIP, PQ_SLSM, PQ_SPRAY, "nada");
```
Comment (schuay) [inline src/bench/random.cpp]: What's the point of the last "nada" entry?

File: src/bench/random.cpp [hunk 5]
```diff
@@ -378,12 +391,13 @@ bench_thread(PriorityQueue *pq,
                              settings.size - thread_id * slice_size : slice_size;
     for (int i = 0; i < initial_size; i++) {
         uint32_t elem = keygen.next();
+        uint32_t value = keygen.next();
```
Comment (schuay) [inline src/bench/random.cpp]: Is it beneficial to add a separate value? I reused elem since the value seems irrelevant and it just adds keygen overhead. Same below.

File: src/bench/pqs/linden.cpp [hunk 6]
```diff
@@ -35,16 +36,43 @@ Linden::~Linden()
...
 
 bool
 Linden::delete_min(uint32_t &v)
-{
-    v = deletemin(m_q->pq);
-    return true;
+{   
+    unsigned long k_ret;
+    v = (uint32_t)deletemin_key(m_q->pq, &k_ret);
+    if(k_ret == ((unsigned long)-1)){
```
Comment (schuay) [inline src/bench/pqs/linden.cpp]: Nit: return (k_ret != -1)

File: src/bench/pqs/linden.cpp [hunk 7]
```diff
@@ -35,16 +36,43 @@ Linden::~Linden()
...
+    }
 }
 
+bool
+Linden::delete_min(size_t &k, size_t &v)
+{   
+    unsigned long k_ret;
+    v = deletemin_key(m_q->pq, &k_ret);
+    k = k_ret -1;
+    if(k_ret == ((unsigned long)-1)){
```
Comment (schuay) [inline src/bench/pqs/linden.cpp]: Same here.

File: lib/spraylist_linden/ssalloc.c [hunk 8]
```diff
@@ -30,14 +30,20 @@ ssalloc_set(void* mem)
...
+
 void
 ssalloc_init()
 {
 #if !defined(SSALLOC_USE_MALLOC)
   int i;
   for (i = 0; i < SSALLOC_NUM_ALLOCATORS; i++)
     {
+#ifdef SSALLOC_SIZE_ALL
+      ssalloc_app_mem[i] = (void*) memalign(64, SSALLOC_SIZE_ALL / number_of_threads);
```
Comment (schuay) [inline lib/spraylist_linden/ssalloc.c]: Just wondering, why can't we pass this into ssalloc_init(int number_of_threads) instead of using a global variable?

File: src/k_lsm/k_lsm_inl.h [hunk 9]
```diff
@@ -82,6 +82,55 @@ k_lsm<K, V, Rlx>::insert(const K &key,
 
 template <class K, class V, int Rlx>
 bool
+k_lsm<K, V, Rlx>::delete_min(K &key, V &val)
```
Comment (schuay) [inline src/k_lsm/k_lsm_inl.h]: Please don't duplicate all of this code. Why not implement delete_min(val) as 

delete_min(V &val) {
  K key;
  return delete_min(key, val);
}

Same in all other files of this commit.

Comment (schuay) [reply src/k_lsm/k_lsm_inl.h]: Also, can you split the delete_min changes into a separate CL?

**Review by schuay — COMMENTED** [review]
Awesome! Looks good to me in general, with a few minor comments inline.

File: bench.py (hunk 1 above)
Comment (kjellwinblad) [reply bench.py]: Will fix.

File: src/bench/random.cpp (hunk 3 above)
Comment (kjellwinblad) [reply src/bench/random.cpp]: Will fix.

File: src/bench/random.cpp (hunk 4 above)
Comment (kjellwinblad) [reply src/bench/random.cpp]: My mistake. I will remove it.

File: src/bench/pqs/linden.cpp (hunk 7 above)
Comment (kjellwinblad) [reply src/bench/pqs/linden.cpp]: Will fix.

File: lib/spraylist_linden/ssalloc.c (hunk 8 above)
Comment (kjellwinblad) [reply lib/spraylist_linden/ssalloc.c]: I don't think there is much reason not to do that. I will try to remove the global variable.

File: src/k_lsm/k_lsm_inl.h (hunk 9 above)
Comment (kjellwinblad) [reply src/k_lsm/k_lsm_inl.h]: I agree. There is no reason to duplicate all that code. Will fix that and split out the delete_min changes to a separate commit.

File: lib/capq/binheap_test.c (hunk 2 above)
Comment (kjellwinblad) [reply lib/capq/binheap_test.c:23]: I don't think it hurts to keep it where it is but I can remove it or move it if you insist?

**Comment by kjellwinblad** [comment]
Thank you for the feedback! I will update the pull request as soon as I have got enough time to do the changes.

File: src/bench/random.cpp (hunk 5 above)
Comment (kjellwinblad) [reply src/bench/random.cpp]: My mistake again. I think I added a separate value to test something and then I must have forgotten to change it back.

File: src/bench/pqs/linden.cpp (hunk 6 above)
Comment (kjellwinblad) [reply src/bench/pqs/linden.cpp]: Will fix.

File: lib/capq/binheap_test.c (hunk 2 above)
Comment (schuay) [reply lib/capq/binheap_test.c:23]: It's fine, up to you. I don't care too much about stuff in lib/ :)

**Comment by kjellwinblad** [comment]
I have made changes according to your feedback. Please, check that everything looks alright and let me know if you want me to fix something more.

File: src/bench/pqs/multiq_inl.h [hunk 10]
```diff
@@ -36,13 +36,13 @@ multiq<K, V, C>::~multiq()
 
 template <class K, class V, int C>
 bool
-multiq<K, V, C>::delete_min(V &value)
+    multiq<K, V, C>::delete_min(K &key, V &value)
```
Comment (schuay) [inline src/bench/pqs/multiq_inl.h]: Nit: Whitespace.

File: src/bench/pqs/multiq_inl.h [hunk 11]
```diff
@@ -57,11 +57,17 @@ multiq<K, V, C>::delete_min(V &value)
         const auto item = pq.top();
 
         if (item.key == SENTINEL_KEY) {
-            // Empty queue, retry indefinitely.
-            // TODO: Not a permanent solution, but original data structure does the same.
+            // Empty queue, retry a few times.
+            // TODO: Not a permanent solution
             unlock(i);
-            return false;
+            if(retry_count == 0){
```
Comment (schuay) [inline src/bench/pqs/multiq_inl.h]: Nit: Please preserve the whitespace style of the remaining code.

File: src/bench/pqs/multiq_inl.h [hunk 12]
```diff
@@ -57,11 +57,17 @@ multiq<K, V, C>::delete_min(V &value)
...
+                return false;
+            }else{
+                retry_count--;
+                goto delete_min_start_label;
```
Comment (schuay) [inline src/bench/pqs/multiq_inl.h]: We can remove the label and just do a 'continue'.

File: src/bench/pqs/linden.cpp [hunk 13]
```diff
@@ -35,16 +36,38 @@ Linden::~Linden()
...
-               const uint32_t & /* Unused */)
+               const uint32_t &value)
 {
-    linden_insert(m_q->pq, key);
+    /*Add 1 to key to support 0 keys*/
+    linden_insert(m_q->pq, key+1, value);
+}
+
+void
+Linden::insert(const size_t &key,
```
Comment (schuay) [inline src/bench/pqs/linden.cpp:46]: Nit: Is this overload really needed? Same for all other sites. Can't we instead make sure that call-sites all use uint32_t's?

File: src/bench/pqs/spraylist.cpp [hunk 14]
```diff
@@ -55,14 +55,38 @@ void
...
-spraylist::delete_min(uint32_t &v)
+spraylist::delete_min(size_t &k, size_t &v)
 {
-    const int ret = spray_delete_min(m_q, &v, d);
-    return (ret == 1);
+    unsigned long k_ret;
+    unsigned long v_ret;
+    int ret;
+    do {
+        ret = spray_delete_min_key(m_q, &k_ret, &v_ret, d);
```
Comment (schuay) [inline src/bench/pqs/spraylist.cpp:75]: Why did this change? I couldn't find any documentation about this in the spraylist code above. Please add a comment explaining the condition below, assuming it's the right thing to do.

**Review by schuay — COMMENTED** [review]
Thanks, looks almost ready to land. I have some final comments.

File: src/bench/pqs/spraylist.cpp (hunk 14 above)
Comment (kjellwinblad) [reply src/bench/pqs/spraylist.cpp:75]: It requires some detective work to figure out that this is how one is supposed to use spray_delete_min_key. I figured it out from the "sssp.c" benchmark in the SprayList repository *. My understanding is that spray_delete_min_key can sometimes fail even though the SprayList is not empty, in which case spray_delete_min_key returns 0 and sets the key to a value that is different from -1. I agree that we should document why we are using spray_delete_min_key in this way. Can you first check that you agree that what I did seems correct?

@jkopinsky , can you also check that spray_delete_min_key is used in the correct way here?

* https://github.com/kjellwinblad/klsm/blob/1ca3340a752dcf57400a9a9b67e84b2ae8bacf94/lib/spraylist_linden/sssp.c#L158

File: src/bench/pqs/linden.cpp (hunk 13 above)
Comment (kjellwinblad) [reply src/bench/pqs/linden.cpp:46]: I think we could make sure that call sites use uint32_t but that would require casting from size_t or other changes so I don't think it is a good idea. Do you have any particular reason for wanting to avoid the overloads?

Comment (schuay) [reply src/bench/pqs/linden.cpp:46]: I just find it unnecessary code duplication. Wouldn't it be an option to settle on 32- or 64-bit ints for use in benchmarks? Another idea would be to use templated insert() and delete_min() methods?

File: src/bench/pqs/spraylist.cpp (hunk 14 above)
Comment (schuay) [reply src/bench/pqs/spraylist.cpp:75]: I see, thanks for the explanation. I'm not sure it's a good idea to loop within delete_min though. For instance, the klsm can also fail spuriously but that is part of the contract for delete_min.

I think handling these cases (potentially with loops) should be done outside delete_min. Not sure how the SprayList works internally though, so I may be wrong.

File: src/bench/pqs/linden.cpp (hunk 13 above)
Comment (kjellwinblad) [reply src/bench/pqs/linden.cpp:46]: I think it is an excellent idea to settle on 64-bit keys and values. I think 64-bit would be better than 32-bit as:

* that is what the priority queues that are implemented in C (Linden, SprayList and CA-PQ) are hard coded for on modern platforms and
* it would make it possible to fit a memory address in the value field.

However, I think such a change is not related to this pull request so perhaps it is better to do this as a separate commit once this pull request has been accepted. What do you think?

File: src/bench/pqs/spraylist.cpp (hunk 14 above)
Comment (kjellwinblad) [reply src/bench/pqs/spraylist.cpp:75]: I did it this way because it makes the file_shortest_paths benchmark (src/bench/file_shortest_paths.cpp) together with the SprayList behave in the same way as the shortest paths benchmark implemented by @jkopinsky (lib/spraylist_linden/sssp.c). I don't see any other good way to accomplish this but I'm open for suggestions.

If I'm not mistaken, the file_shortest_paths benchmark used with the SprayList could compute the wrong result if this loop was not present even though it is probably very unlikely in practice. The reason is that the spurious failures seems to depend on a random number generator so if all worker threads spuriously fail enough times at the wrong time the algorithm will stop even though there is still work to do. Again if I'm not mistaken, the file_shortest_paths benchmark together with KLSM will always compute the correct result even though KLSM can spuriously fail. The reason is that all worker threads can't fail "at the same time" with KSLM. The stop condition for the file_shortest_paths benchmark is that all worker threads called delete_min after some time point after which no insert operations have been issued and got false as return value.

Comment (jkopinsky) [reply src/bench/pqs/spraylist.cpp:75]: Yes, currently SprayList can fail spuriously either due to failed Compare-and-Swap (unlikely by design) or due to randomness even if the list is non-empty (also unlikely, though chance of this increases if the list is very small). 

I think it was a planned feature to have retries happen internally but this never got implemented; sorry for the confusion.

File: src/bench/pqs/linden.cpp (hunk 13 above)
Comment (schuay) [reply src/bench/pqs/linden.cpp:46]: Ok, let's do that.

File: src/bench/pqs/spraylist.cpp (hunk 14 above)
Comment (schuay) [reply src/bench/pqs/spraylist.cpp:75]: The src/bench/shortest_paths.cpp benchmark has a stop condition based on a global counter of the number of active tasks, so I think it calculates the correct result regardless of spurious failures - is that not the case? See https://github.com/klsmpq/klsm/blob/master/src/bench/shortest_paths.cpp#L192, a failure just results in retrying the deletion.

Anyway, if a non-spurious failure mode does turn out to be required, maybe we should add it as another method to avoid influencing existing benchmarks.

Once this question is resolved, it's all good to land from my side :+1:

Comment (kjellwinblad) [reply src/bench/pqs/spraylist.cpp:75]: @schuay Yes, I think you are right, src/bench/shortest_paths.cpp calculates the correct result regardless of spurious failures. Unfortunately, the global counter is a serious scalability problem. That is why src/bench/file_shortest_paths.cpp does not depend on a global counter. Instead src/bench/file_shortest_paths.cpp use a stop condition that grantee correctness as long as the priority queue has the property described in Theorem 4  in the paper  [The Contention Avoiding Concurrent Priority Queue LCPC'2016](http://www.it.uu.se/research/group/languages/software/ca_pq/relax_prio_queue_paper.pdf). I'm quite sure that KLSM satisfies this property and I don't see any reason why SprayList shouldn't.

I was surprised to see that there seems to be a small difference between using the retry loop and not using the retry loop as retries should be rare (see measurements below). Anyway, my opinion is that the proper version to measure is the one with the retry loop. I interpret @jkopinsky's answer above as that this is also his opinion. The only reason I see to expose SprayList's spurious failures to the application is to allow the application to add a custom yield or sleep before the retry. Changing "delete_min(uint32_t &v)" to look as the code that can be found below will avoid influencing the existing benchmarks. Do you want me to do that?

## SprayList Measurements
The measurements where done using the version with commit id 1ca3340a752dcf57400a9a9b67e84b2ae8bacf94. I ran it on a "AMD FX(tm)-6100 Six-Core Processor". The "retry loop version" has the retry loop and the "without retry loop" version has a delete_min method modified to look like this:

```
bool
spraylist::delete_min(uint32_t &v)
{
    size_t k_ret;
    size_t v_ret;
    const int ret = spray_delete_min_key(m_q, &k_ret, &v_ret, d);
    v = (uint32_t)v_ret;
    return ret == 1;
} 
```


### retry loop version 
```
./build/src/bench/random -p 6 spray
3890564
$ ./build/src/bench/random -p 6 spray
3893154
$ ./build/src/bench/random -p 6 spray
3902697
```
### without retry loop
```
$ ./build/src/bench/random -p 6 spray
4049361
$ ./build/src/bench/random -p 6 spray
4063919
$ ./build/src/bench/random -p 6 spray
4049860
```

Comment (schuay) [reply src/bench/pqs/spraylist.cpp:75]: Wow that's a pretty noticeable difference. Anyway, I'll defer to your best judgement on this since it's  been a while that I looked at all of this.

My gut reaction is that 1. it'd be nice to keep the behavior of existing benchmarks consistent (unless they're incorrect), 2. behavior of different PQs should be consistent (it seems weird if only some loop in delete_min), and 3. making the shortest path benchmark more scalable sounds good.

Let me know when you think this is ready to merge.

Comment (kjellwinblad) [reply src/bench/pqs/spraylist.cpp:75]: I will try to give answers to the first two of your gut reactions (please let me know if you don't agree with something):

1. I'm not sure if the old measurements in the random benchmark with the spraylist are incorrect given that you also measure the quality of the delete_min results. No retry loop in delete_min should give better performance but worse quality and the other way around. However, I think the chance that the results have relevance for a real application is larger with the retry loop as that is what one would like to have in a real application.
2. I think the reason you think it is weird to only have the retry loop for SparyList is that the retry loop should actually be inside SprayList's "own" delete_min function. I bet KLSM has some retry loop internally that instead could be handled by spurious failures, right? The spurious failures that the retry loop retries on are quite different from KLSM's spurious failures (I think) in how easily detectable they are.

Comment (schuay) [reply src/bench/pqs/spraylist.cpp:75]: Re 2.: The KLSM doesn't have an internal retry loop, it only retries once after spying if the first attempt failed IIRC. It just seems weird to me that some delete_min's have a retry loop and others don't.

Comment (kjellwinblad) [reply src/bench/pqs/spraylist.cpp:75]: The KLSM as decribed in the [The Lock-free k-LSM Relaxed Priority Queue](https://arxiv.org/abs/1503.05698) paper does have a retry loop (See lines 15-25 in Listing 5). Doesn't the actual implementation have the corresponding loop? I totally agree that it is weird to have it for the SprayList and not for the KLSM, if the KLSM doesn't have this loop.

Comment (schuay) [reply src/bench/pqs/spraylist.cpp:75]: You're right, there is a retry loop here: https://github.com/klsmpq/klsm/blob/master/src/k_lsm/k_lsm_inl.h#L104

It retries if both the local dlsm and the shared slsm are empty, and spying (taking items from another thread's local dlsm) copied more than 0 items to the local dlsm.

It differs from Listing 5 in that failed CAS result in a spurious failure.

I'm fine with keeping the SprayList loop if you think that is appropriate (it'd be great to have a short comment as to what the conditions (return value, key value) mean).

Comment (kjellwinblad) [reply src/bench/pqs/spraylist.cpp:75]: > It differs from Listing 5 in that failed CAS result in a spurious failure.

Why doesn't the implementation do as the paper? I think that it would make more sense if the implementation retried on CAS failure as the code in the paper does. I think so for the same reasons as I think the retry loop that I added to SprayList's delete_min should be there.

I was wrong in thinking that the KLSM implementation didn't have this kind of spurious failures. I thought the KLSM implementation's spurious failures happened when the issuing thread's local queue and the shared queue both were empty but some other local queue had items, which I could understand as it would be costly to detect this kind of spurious failures.

Anyway, if it is decided to change the KLSM, I think is better that you @schuay do it since you are more familiar with the KLSM code. I will go ahead and do the other changes that you have suggested so we can finish this pull request.

Comment (schuay) [reply src/bench/pqs/spraylist.cpp:75]: That paper was published before this standalone version existed, and describes the implementation in the task scheduling framework Pheet (https://launchpad.net/pheet), which also evolved over time. Implementation differences are expected.

Anyway, I agree that we should proceed with landing this PR soon.

**Comment by kjellwinblad** [comment]
Just made the changes that you suggested to the multi queue. Thanks for the feedback. I think it should be ready to merge now unless you find anything else.

**Comment by schuay** [comment]
Thanks, great to have this!
